src/sentry/services/hybrid_cloud/audit/impl.py

```diff
@@ -0,0 +1,68 @@
+from __future__ import annotations
+
+from typing import Any, Mapping
+
+from sentry.models import AuditLogEntry, AuthIdentity, AuthProvider, Organization, Team, User
+from sentry.services.hybrid_cloud.audit import AuditLogMetadata, AuditLogService
+from sentry.services.hybrid_cloud.auth import ApiAuthIdentity, ApiAuthProvider
+from sentry.services.hybrid_cloud.organization import ApiOrganizationMember
+from sentry.services.hybrid_cloud.user import APIUser, user_service
+
+
+class DatabaseBackedAuditLogService(AuditLogService):
+    @staticmethod
+    def _resolve_user(user: APIUser | int | None) -> User | None:
+        if user is None:
+            return None
+        if isinstance(user, int):
+            user_id = user
+        else:
+            user_id = user.id
+        return User.objects.get(id=user_id)
+
+    def write_audit_log(self, *, metadata: AuditLogMetadata, data: Mapping[str, Any]) -> None:
+        organization = Organization.objects.get(id=metadata.organization.id)
+        AuditLogEntry.objects.create(
+            organization=organization,
+            event=metadata.event,
+            actor_label=metadata.actor_label,
+            actor=self._resolve_user(metadata.actor),
+            ip_address=metadata.ip_address,
+            target_object=metadata.target_object,
+            target_user=self._resolve_user(metadata.target_user),
+            data=data,
+        )
+
+    def log_organization_membership(
+        self, *, metadata: AuditLogMetadata, organization_member: ApiOrganizationMember
+    ) -> None:
+        user_id = organization_member.user_id
+        assert user_id is not None
```

Comment:
Hmmm... I think this is wrong.  Just because a user does not exist, does not mean we cannot log something about their id.  Many AuditLogEntry invocations come from async processes, where technically they can have a user_id in hand of a user that has already been deleted.  At the moment, we use cascade logic to `set_null` those user_ids, but in the future an async process will provide that. 

I think the correct thing is to take the user_id provided "as is" without validating its deletion and allow eventually consistency of the tombstone / outbox pattern I'm working on to handle the rest.

So basically, *we still want audit log entries even when a user no longer exists*, and *we'll handle the set null at a different level*.

Reply:
Sounds good. I would guess that, when I dropped in that assertion, I was thinking more about getting Mypy to pass than about the correct behavior for audit logs.

It also wouldn't surprise me if, before my changes, the rare case of a null user would have simply caused an error, and it's just more obvious now.